types/src/v0/impls/stake_table.rs

```diff
@@ -239,168 +215,220 @@ pub fn validators_from_l1_events<I: Iterator<Item = StakeTableEvent>>(
+                    ));
+                }
+
+                let entry = self.validators.entry(account);
+                if let indexmap::map::Entry::Occupied(_) = entry {
+                    return Err(StakeTableError::AlreadyRegistered(account));
+                }
 
-                bls_keys.insert(stake_table_key);
-                schnorr_keys.insert(state_ver_key.clone());
+                entry.or_insert(Validator {
+                    account,
+                    stake_table_key,
+                    state_ver_key,
+                    stake: U256::ZERO,
+                    commission,
+                    delegators: HashMap::new(),
+                });
 
-                match validators.entry(account) {
-                    indexmap::map::Entry::Occupied(_occupied_entry) => {
-                        bail!("validator {:#x} already registered", *account)
-                    },
-                    indexmap::map::Entry::Vacant(vacant_entry) => vacant_entry.insert(Validator {
-                        account,
-                        stake_table_key,
-                        state_ver_key,
-                        stake: U256::from(0_u64),
-                        commission,
-                        delegators: HashMap::default(),
-                    }),
-                };
+                Ok(())
             },
+
             StakeTableEvent::Deregister(exit) => {
-                validators
+                self.validators
                     .shift_remove(&exit.validator)
-                    .with_context(|| format!("validator {:#x} not found", exit.validator))?;
+                    .ok_or(StakeTableError::ValidatorNotFound(exit.validator))?;
+
+                Ok(())
             },
+
             StakeTableEvent::Delegate(delegated) => {
                 let Delegated {
                     delegator,
                     validator,
                     amount,
                 } = delegated;
-                let validator_entry = validators
+
+                let val = self
+                    .validators
                     .get_mut(&validator)
-                    .with_context(|| format!("validator {validator:#x} not found"))?;
+                    .ok_or(StakeTableError::ValidatorNotFound(validator))?;
 
                 if amount.is_zero() {
                     tracing::warn!("delegator {delegator:?} has 0 stake");
-                    continue;
+                    return Ok(());
                 }
-                // Increase stake
-                validator_entry.stake += amount;
+
+                val.stake += amount;
                 // Insert the delegator with the given stake
                 // or increase the stake if already present
-                validator_entry
-                    .delegators
+                val.delegators
                     .entry(delegator)
-                    .and_modify(|stake| *stake += amount)
+                    .and_modify(|s| *s += amount)
```

Comment:
`stake` is too many character? :D

Reply:
reverted https://github.com/EspressoSystems/espresso-network/pull/3396/commits/ae5f382583faf8c20dc91336be5ffe1db95cc0f3